nipap/nipap/authlib.py

```diff
@@ -327,17 +335,45 @@ def authenticate(self):
 
         # auth succeeded
         self.authenticated_as = self.username
-        self._authenticated = True
         self.trusted = False
         self.readonly = False
 
         try:
-            res = self._ldap_conn.search_s(self._ldap_basedn, ldap.SCOPE_SUBTREE, 'uid=' + self.username, ['cn'])
+            res = self._ldap_conn.search_s(self._ldap_basedn, ldap.SCOPE_SUBTREE, self._ldap_search.format(ldap.dn.escape_dn_chars(self.username)), ['cn','memberOf'])
             self.full_name = res[0][1]['cn'][0]
+            # check for ro_group membership if ro_group is configured
+            if self._ldap_ro_group:
+                if self._ldap_ro_group in res[0][1]['memberOf']:
+                    self.readonly = True
+            # check for rw_group membership if rw_group is configured
+            if self._ldap_rw_group:
+                if self._ldap_rw_group in res[0][1]['memberOf']:
+                    self.readonly = False
+                else:
+                    # if ro_group is configured, and the user is a member of
+                    # neither the ro_group nor the rw_group, fail authentication.
+                    if self._ldap_ro_group:
+                        if self._ldap_ro_group not in res[0][1]['memberOf']:
+                            self._authenticated = False
+                            return self._authenticated
+                    else:
+                        self.readonly = True
+
         except:
             self.full_name = ''
+            # authentication fails if both ro_group and rw_group are configured
+            # and the search fails.
+            if self._ldap_rw_group and self._ldap_ro_group:
+                self._authenticated = False
+                return self._authenticated
+            # if either ro_group or rw_group are configured and the search
+            # fails, give readonly access
+            elif self._ldap_rw_group or self._ldap_ro_group:
```

Comment:
I'd agree with giving read only rather than read/write access here.

However, this except clause will only be reached if search_s fails, which should never happen, except when it's passed an invalid filter string (_ldap_search is misconfigured), or when there's a problem communicating with the LDAP server (e.g. connection has dropped since binding). The former is likely to be much more common.

Maybe it would be better to raise an exception instead?